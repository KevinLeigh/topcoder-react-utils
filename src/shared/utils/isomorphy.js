/**
 * Collection of helpers to deal with isomorphic aspects of the code.
 */

/**
 * Returns `true` when executed at the front end side; `false` otherwise.
 * @return {Boolean}
 */
export function isClientSide() {
  return Boolean(global.TRU_FRONT_END);
}

/**
 * Returns `true` if executed at the server side; `false` otherwise.
 * @return {Boolean}
 */
export function isServerSide() {
  return !global.TRU_FRONT_END;
}

/**
 * @return {String} Code mode: "development" or "production".
 */
function getMode() {
  return isClientSide() ? global.TRU_CONSTANTS.mode : process.env.BABEL_ENV;
}

/**
 * Returns `true` if development version of the code is running;
 * `false` otherwise.
 * @return {Boolean}
 */
export function isDevBuild() {
  return getMode() === 'development';
}

/**
 * Returns `true` if production build of the code is running;
 * `false` otherwise.
 * @return {Boolean}
 */
export function isProdBuild() {
  return getMode() === 'production';
}

/**
 * Returns build timestamp of the front-end JS bundle.
 * @return {String} ISO date/time string.
 */
export function buildTimestamp() {
  return global.TRU_CONSTANTS.timestamp;
}
