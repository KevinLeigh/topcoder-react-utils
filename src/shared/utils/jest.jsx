/**
 * Utilities for Jest tests.
 */

/* NOTE: At the moment here is no intention to use this module for runtime
 * execution, only for tests; thus we depend on Jest environment and dev
 * dependencies. */

/* global expect */

import PT from 'prop-types';
import React from 'react';
import TU from 'react-dom/test-utils';

/* eslint-disable import/no-extraneous-dependencies */
import Renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
/* eslint-enable import/no-extraneous-dependencies */

/**
 * Just an alias for TU.findRenderedDOMComponentWithClass(..).
 * @param {Object} dom
 * @param {String} className
 * @return {Object}
 */
export function findInDomByClass(dom, className) {
  return TU.findRenderedDOMComponentWithClass(dom, className);
}

/**
 * Just an alias for TU.scryRenderedDOMComponentsWithClass(..).
 * @param {Object} dom
 * @param {Stirng} className
 * @return {Array}
 */
export function findInDomManyByClass(dom, className) {
  return TU.scryRenderedDOMComponentsWithClass(dom, className);
}

/**
 * Auxiliary wrapper around ReactJS Test Renderer.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of the rendered tree.
 */
export function render(component) {
  return Renderer.create(component).toJSON();
}

/* The Wrapper is necessary for the "renderDom(..)" function, because
 * the "renderIntoDocument(..)" function from "react-dom/test-utils" works
 * only with state components, so we have to wrap our ReactJS components into
 * such Wrapper. */
class Wrapper extends React.Component {
  componentDidMount() {}

  render() {
    return this.props.children;
  }
}

Wrapper.propTypes = {
  children: PT.node.isRequired,
};

/**
 * Renders given component into DOM, using "react-dom/test-utils".
 * @param {Object} component ReactJS component to render.
 * @return {Object} Rendered DOM.
 */
export function renderDom(component) {
  return TU.renderIntoDocument(<Wrapper>{component}</Wrapper>);
}

/**
 * Auxiliary wrapper around ReactJS Shallow Test Renderer.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of the shallow component's render tree.
 */
export function shallowRender(component) {
  const renderer = new ShallowRenderer();
  renderer.render(component);
  return renderer.getRenderOutput();
}

/**
 * Renders provided ReactJS component with ReactJS Shallow Test Renderer,
 * creates/tests its snapshot, and returns the render result.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of shallow render.
 */
export function shallowSnapshot(component) {
  const res = shallowRender(component);
  expect(res).toMatchSnapshot();
  return res;
}

/**
 * Renders provided ReactJS component with ReactJS Test Renderer,
 * creates/tests the snapshot, and returns the render.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON render of the component.
 */
export function snapshot(component) {
  const res = render(component);
  expect(res).toMatchSnapshot();
  return res;
}

export const simulate = TU.Simulate;
