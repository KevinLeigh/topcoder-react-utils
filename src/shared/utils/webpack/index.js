/**
 * Requires the specified module without including it into the bundle during
 * Webpack build. This function should be executed only server-side.
 * @param {String} modulePath
 * @return Required module.
 */
export function requireWeak(modulePath) {
  /* eslint-disable global-require, import/no-dynamic-require */
  const mod = require('./require')(modulePath);
  /* eslint-enable global-require, import/no-dynamic-require */
  return mod.default || mod;
}

export default undefined;
