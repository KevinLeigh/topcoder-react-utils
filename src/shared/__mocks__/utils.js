import * as mock from '../utils';
import isomorphy from '../utils/__mocks__/isomorphy';

mock.isomorphy = isomorphy;

export default mock;
