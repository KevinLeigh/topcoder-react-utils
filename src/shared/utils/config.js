/**
 * This module provides isomorphic configuration. At the client-side it serves
 * cofiguration object from the global CONFIG variable, injected into the page
 * during the server-side rendering; at the server side it is fetched directly
 * from node-config.
 */

/* eslint-env browser */

import { isClientSide } from './isomorphy';

/* eslint-disable global-require */
module.exports = isClientSide() ? window.CONFIG : require('config');
/* eslint-enable global-require */
