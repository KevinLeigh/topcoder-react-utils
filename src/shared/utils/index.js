import * as isomorphy from './isomorphy';

export default {
  isomorphy,
};
