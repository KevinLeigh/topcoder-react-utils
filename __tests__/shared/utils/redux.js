import _ from 'lodash';
import * as redux from 'utils/redux';
import { createAction } from 'redux-actions';

const DUMMY_ACTION = {
  type: 'DUMMY/ACTION',
  error: true,
  payload: 'DUMMY_PAYLOAD',
};

const sampleActionCreator = createAction('SAMPLE/ACTION', (...args) => args);

test('Module exports', () => expect(redux).toMatchSnapshot());

test('proxyAction: payload creator reconstruction', () => {
  const proxy = redux.proxyAction(sampleActionCreator);
  expect(proxy('ARG1', 'ARG2')).toEqual(['ARG1', 'ARG2']);
});

test('proxyAction: action mapping', () => {
  expect(redux.proxyAction(sampleActionCreator, _.clone(DUMMY_ACTION)))
    .toEqual({ ...DUMMY_ACTION, type: 'SAMPLE/ACTION' });
});

test('proxyReducer', () => {
  const mockReducer = jest.fn();
  const proxyReducer = redux.proxyReducer(mockReducer, sampleActionCreator);
  proxyReducer('DUMMY_STATE', _.clone(DUMMY_ACTION));
  expect(mockReducer).toHaveBeenCalledWith('DUMMY_STATE', {
    ...DUMMY_ACTION,
    type: 'SAMPLE/ACTION',
  });
});
