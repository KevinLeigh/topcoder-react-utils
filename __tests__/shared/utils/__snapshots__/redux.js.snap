// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module exports 1`] = `
Object {
  "combineReducers": [Function],
  "connect": [Function],
  "createActions": [Function],
  "handleActions": [Function],
  "proxyAction": [Function],
  "proxyReducer": [Function],
  "resolveAction": [Function],
  "resolveReducers": [Function],
  "storeFactory": [Function],
}
`;
